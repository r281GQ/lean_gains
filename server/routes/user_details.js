const {
  handleGetUserDetails,
  handlePutUserDetails
} = require('./../services/user_details');

module.exports = app => {
  app.get('/api/userdetails', (request, response) => {
    handleGetUserDetails(request, response)
      .then(userDetails => response.status(200).send(userDetails))
      .catch(error => console.log(error));
  });
  
  app.put('/api/userdetails', (request, response) => {
    handlePutUserDetails(request, response)
      .then(userDetails => response.status(200).send(userDetails))
      .catch(error => console.log(error));
  });
};
