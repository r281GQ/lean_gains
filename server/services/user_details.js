const mongoose = require('mongoose');
const moment = require('moment');
const _ = require('lodash');
const User = mongoose.model('User');

const handleGetUserDetails = (request, response) => {
  return new Promise((resolve, reject) => {
    User.findOne({_id: request.user._id})
    .then(user => {
      const details = _.pick(user, ['dob', 'sex', 'picture', 'userName']);
      resolve(details)
    })
    .catch(error => reject(error))
  })
}



const handlePutUserDetails = (request, response) => {
  const {dob, sex, userName} =  request.body;
  return new Promise((resolve, reject) => {
    User.findOneAndUpdate({_id: request.user._id},
      {$set: {dob, sex, userName}}, {new: true})
    .then(user => {
      const details = _.pick(user, ['dob', 'sex', 'picture', 'userName']);

      resolve(details)
    })
    .catch(error => reject(error))
  })
}

module.exports = {
  handleGetUserDetails,handlePutUserDetails
}
