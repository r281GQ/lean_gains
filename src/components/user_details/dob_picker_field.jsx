import React from 'react';
import { Field } from 'redux-form/immutable';
import { DatePicker } from 'redux-form-material-ui';
import moment from 'moment';

const DOBPickerField = ({ minDate, maxDate }) => {
  return (
    <Field
      name="dob"
      component={input =>
        <DatePicker
          {...input}
          maxDate={maxDate}
          minDate={minDate}
          formatDate={value => moment(value).format('DD-MM-YYYY')}
        />}
    />
  );
};

export default DOBPickerField;
