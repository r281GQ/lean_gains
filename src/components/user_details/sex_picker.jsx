import React from 'react';
import { Field } from 'redux-form/immutable';
import { RadioButtonGroup } from 'redux-form-material-ui';
import { RadioButton } from 'material-ui';

const SexPicker = () =>
  <Field name="sex" component={RadioButtonGroup}>
    <RadioButton value="male" label="Male" />
    <RadioButton value="female" label="Female" />
  </Field>;

export default SexPicker;
