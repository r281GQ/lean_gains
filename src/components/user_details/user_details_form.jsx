import React from 'react';
import PropTypes from 'prop-types';
import { Paper } from 'material-ui';

import UserNamePicker from './user_name_picker';
import SubmitButton from './../submit_button';
import SexPicker from './sex_picker';
import DOBPickerField from './dob_picker_field';

const UserDetailsForm = ({
  handleUpdateUserDetails,
  handleSubmit,
  validators,
  normalizeDate
}) =>
  <Paper className="user-details-paper">
    <form onSubmit={handleUpdateUserDetails}>
      <UserNamePicker validator={validators.userName} />
      <DOBPickerField
        minDate={validators.minDate}
        maxDate={validators.maxDate}
        normalize={normalizeDate}
      />
      <SexPicker />
      <SubmitButton label={`Update`} />
    </form>
  </Paper>;

UserDetailsForm.propTypes = {
  handleUpdateUserDetails: PropTypes.func.isRequired,
  normalizeDate: PropTypes.func.isRequired,
  validators: PropTypes.shape({
    userName: PropTypes.func.isRequired,
    minDate: PropTypes.object.isRequired,
    maxDate: PropTypes.object.isRequired
  })
};

export default UserDetailsForm;
