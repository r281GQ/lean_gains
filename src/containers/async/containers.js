import asyncComponent from './async_container';

export const AsyncMainContainer = asyncComponent(() =>
  System.import('./../main').then(module => module.default)
);

export const AsyncDailyLogsRouter = asyncComponent(() =>
  System.import('./../daily_logs/daily_logs_router').then(
    module => module.default
  )
);

export const AsyncCalorieTargetContainer = asyncComponent(() =>
  System.import('./../calorie_target/calorie_target_container').then(module => module.default)
);

export const AsyncUserDetailsContainer = asyncComponent(() =>
  System.import('./../user_details').then(module => module.default)
);

export const AsyncWorkoutLogsRouter = asyncComponent(() =>
  System.import('./../workout_logs/workout_logs_router').then(
    module => module.default
  )
);

export const AsyncWorkoutTargetsRouter = asyncComponent(() =>
  System.import('./../workout_targets/workout_targets_router').then(
    module => module.default
  )
);

export const AsyncCalorieTrackerContainer = asyncComponent(() =>
  System.import('./../calorie_tracker/calorie_tracker_container').then(module => module.default)
);
