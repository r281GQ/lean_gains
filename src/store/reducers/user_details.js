import Immutable, { fromJS } from "immutable";
import moment from "moment";
import * as _ from "lodash";

import {
  WRITE_USER_DETAILS,
  UNSET_PICTURE,
  WRITE_WORKOUT_TARGETS,
  WRITE_WORKOUT_TARGET,
  DELETE_WORKOUT_TARGET,
  WRITE_KCAL_TARGETS,
  WRITE_LATEST
} from "./../actions/user_details_actions";

export const INITIAL_STATE = Immutable.fromJS({
  _id: undefined,
  dob: undefined,
  gender: undefined,
  picture: undefined,
  email: undefined,
  username: undefined,
  weightDisplayPreference: undefined,
  lengthDisplayPreference: undefined,
  supplements: [],
  workoutTargets: {},
  kcalTargets: {},
  latestMeasurements: {
    weight: 123,
    neck: undefined,
    chest: undefined,
    rightArm: undefined,
    leftArm: undefined,
    aboveBelly: undefined,
    belly: undefined,
    belowBelly: undefined,
    hips: 45,
    rightThigh: undefined,
    leftThigh: undefined
  }
});

const handleWriteUserDetails = (
  state,
  { name, dob, gender, picture, username }
) => {
  return state.withMutations(state =>
    state
      .set("picture", picture)
      .set("username", username)
      .set("name", name)
      .set("dob", moment(dob, "DD-MM-YYYY").valueOf())
      .set("gender", gender)
  );
};

const handleWriteKcalTargets = (state, payload) => {
  payload = _.keyBy(payload, "_id");
  console.log("from reducer", payload);
  return state.withMutations(map =>
    map.delete("kcalTargets").set("kcalTargets", fromJS(payload))
  );
};

const handleWriteWorkoutTargets = (state, payload) => {
  payload = _.keyBy(payload, "_id");
  console.log("from reducer", payload);
  return state.withMutations(map =>
    map.delete("workoutTargets").set("workoutTargets", fromJS(payload))
  );
};

const handleWriteWorkoutTarget = (state, payload) =>
  state.setIn(["workoutTargets", payload._id], fromJS(payload));

const userDetails = (state = INITIAL_STATE, { type, payload }) => {
  switch (type) {
    case WRITE_LATEST:
      return state.set('latestMeasurements', fromJS(payload));
    case WRITE_KCAL_TARGETS:
      return handleWriteKcalTargets(state, payload);
    case DELETE_WORKOUT_TARGET:
      return state.deleteIn(["workoutTargets", payload]);
    case WRITE_WORKOUT_TARGET:
      return handleWriteWorkoutTarget(state, payload);
    case WRITE_WORKOUT_TARGETS:
      return handleWriteWorkoutTargets(state, payload);
    case UNSET_PICTURE:
      return state.set("picture", undefined);
    case WRITE_USER_DETAILS:
      return handleWriteUserDetails(state, payload);
    default:
      return state;
  }
};

export default userDetails;
