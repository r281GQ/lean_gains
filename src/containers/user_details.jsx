import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { reduxForm } from 'redux-form/immutable';
import { connect } from 'react-redux';
import { fromJS } from 'immutable';
import moment from 'moment';

import { updateUserDetails } from './../store/actionCreators/user_details_action_creators';
import { required } from './../services/validators';
import UserDetailsForm from './../components/user_details/user_details_form';

class UserDetailsContainer extends PureComponent {
  componentDidMount() {
    const { userName, dob, sex, initialize } = this.props;
    initialize(fromJS({ userName, dob: moment(dob).toDate(), sex }));
  }

  constructor(props) {
    super(props);
    this._handleFormSubmit = this._handleFormSubmit.bind(this);
    this._normalizeDate = this._normalizeDate.bind(this);
  }

  _handleFormSubmit(formProps) {
    this.props.updateUserDetails({
      userName: formProps.get('userName'),
      sex: formProps.get('sex'),
      dob: moment(formProps.get('dob')).valueOf()
    });
  }

  _normalizeDate() {
    return value => moment(value).valueOf();
  }

  render() {
    const { handleSubmit } = this.props;
    return (
      <div className="user-details">
        <UserDetailsForm
          {...this.props}
          handleUpdateUserDetails={handleSubmit(this._handleFormSubmit)}
          normalizeDate={this._normalizeDate}
          validators={{
            userName: required,
            minDate: moment()
              .subtract(110, 'years')
              .toDate(),
            maxDate: moment()
              .subtract(5, 'years')
              .toDate()
          }}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  sex: state.getIn(['userDetails', 'sex']),
  dob: moment(state.getIn(['userDetails', 'dob'])).toDate(),
  userName: state.getIn(['userDetails', 'userName'])
});

UserDetailsContainer.propTypes = {
  userName: PropTypes.string,
  dob: PropTypes.instanceOf(Date),
  sex: PropTypes.oneOf(['male', 'female']),
  updateUserDetails: PropTypes.func
};

export default connect(mapStateToProps, { updateUserDetails })(
  reduxForm({
    form: 'user-details'
  })(UserDetailsContainer)
);

export { UserDetailsContainer as PureUserDetailsContainer };
