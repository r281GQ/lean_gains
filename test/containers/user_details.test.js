import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { PureUserDetailsContainer } from './../../src/containers/user_details';

describe('Header test', () => {
  it('should render', () => {
    const updateUserDetailsSpy = sinon.spy();
    const dateStub = sinon.stub(Date, 'now').returns(1);
    const component = shallow(
      <PureUserDetailsContainer
        updateUserDetails={updateUserDetailsSpy}
        handleSubmit={() => func => func}
      />
    );
    expect(component.instance()._normalizeDate()).toBeInstanceOf(Function);

    component.instance()._handleFormSubmit({ get: () => undefined });

    expect(updateUserDetailsSpy.calledOnce).toBe(true);

    expect(component).toHaveLength(1);
    expect(component).toMatchSnapshot();
    dateStub.restore();
  });
});
