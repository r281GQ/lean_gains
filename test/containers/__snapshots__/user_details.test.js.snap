// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header test should render 1`] = `
<div
  className="user-details"
>
  <UserDetailsForm
    handleSubmit={[Function]}
    handleUpdateUserDetails={[Function]}
    normalizeDate={[Function]}
    updateUserDetails={[Function]}
    validators={
      Object {
        "maxDate": 1965-01-01T00:00:00.001Z,
        "minDate": 1860-01-01T00:00:00.001Z,
        "userName": [Function],
      }
    }
  />
</div>
`;
